legate/numpy/runtime.py

```diff
@@ -642,72 +628,72 @@ def find_or_create_key_partition(self):
-                tile_offset = np.zeros((len(self.shape),), dtype=np.int64)
-                for n in range(len(self.shape)):
-                    nonzero = False
-                    for m in range(len(root.shape)):
-                        if inverse[n, m] == 0:
-                            continue
-                        nonzero = True
-                        break
-                    if not nonzero:
-                        tile_offset[n] = 1
-                color_lo = tuple((inverse @ color_lo).flatten())
-                color_hi = tuple((inverse @ color_hi).flatten())
-                color_tile = tuple(
-                    (inverse @ color_tile).flatten() + tile_offset
-                )
-                # Reset lo and hi back to our space
-                lo = np.zeros((len(self.shape),), dtype=np.int64)
-                hi = np.array(self.shape, dtype=np.int64) - 1
+                # Shift the tile down (in root space) to align with tile
+                # boundaries, while still covering all of our elements. Then
+                # translate this shift onto view space, as expected by
+                # _find_or_create_partition.
+                offset = tuple(map(lambda x, y: -(x % y), lo, color_tile))
+                offset = tuple((r2v_permute @ offset).flatten())
+                # Translate tile information back to the view space
+                color_lo = r2v_tile.apply(color_lo)
+                color_hi = r2v_tile.apply(color_hi)
+                color_tile = tuple((r2v_permute @ color_tile).flatten())
             else:
-                # If there is no transform then can just use the root function
+                # If there is no transform then just use the root function
                 self.shard_function = rootfn
+                offset = (0,) * len(self.shape)
             self.shard_space = root_key.index_partition.color_space
             # Launch space is how many tiles we have in each dimension
             color_shape = tuple(
                 map(lambda x, y: (x - y) + 1, color_hi, color_lo)
             )
             # Check to see if they are all one, if so then we don't even need
             # to bother with making the partition
-            volume = reduce(lambda x, y: x * y, color_shape)
+            volume = calculate_volume(color_shape)
             assert volume > 0
             if volume == 1:
                 self.launch_space = ()
                 # We overlap with exactly one point in the root
                 # Therefore just record this point
                 self.shard_point = Point(color_lo)
                 return None, None, None
-            # Now compute the offset for the partitioning
-            # This will shift the tile down if necessary to align with the
-            # boundaries at the root while still covering all of our elements
-            offset = tuple(
-                map(
-                    lambda x, y: 0 if (x % y) == 0 else ((x % y) - y),
-                    lo,
```

Comment:
This would always be zeros at this point.

Reply:
That seems strange... Would they be non-zero in the case with added/removed dimensions?

Reply:
No, it was always zeros. This is the slice of the code that updated `lo` in the [original version](https://github.com/nv-legate/legate.numpy/blob/cee64aae695fbfe5cbdd0e22e1e1d7d51bb98e6d/legate/numpy/runtime.py):

```
lo = np.zeros((len(self.shape),), dtype=np.int64)
if self.transform:
  lo = self.transform.apply(lo)
if self.transform:
  lo = np.zeros((len(self.shape),), dtype=np.int64)
offset = f(lo)
```

In every case `lo` would be all zeros at this point.

Reply:
It's possible I was just being overly careful at this point. Are you convinced that's the case, or do you think that there might be other cases and want help trying to find them?

Reply:
You've definitely done more testing than I did and I think you've conceptually got the idea down, so definitely please proceed with this. If you can handle the adding/removing dimensions part I think it will be in good shape.